tracee-rules/engine/engine.go

```diff
@@ -108,17 +124,81 @@ func (engine *Engine) consumeSources(done <-chan bool) {
 						}
 					}
 				}
+				engine.signaturesMutex.RUnlock()
 				engine.inputs.Tracee = nil
 			} else if event != nil {
+				engine.signaturesMutex.RLock()
 				for _, s := range engine.signaturesIndex[types.SignatureEventSelector{Source: "tracee", Name: event.(tracee.Event).EventName}] {
 					engine.signatures[s] <- event
 				}
 				for _, s := range engine.signaturesIndex[types.SignatureEventSelector{Source: "tracee", Name: "*"}] {
 					engine.signatures[s] <- event
 				}
+				engine.signaturesMutex.RUnlock()
 			}
 		case <-done:
 			return
 		}
 	}
 }
+
+//LoadSignature will store in Engine data structures the given signature and activate its handling business logics.
+// It will return the signature ID as well as error.
+func (engine *Engine) LoadSignature(signature types.Signature) (string, error) {
+	selectedEvents, err := signature.GetSelectedEvents()
+	if err != nil {
+		return "", fmt.Errorf("failed to store signature: %w", err)
+	}
+
+	// insert in engine.signatures map
+	metadata, _ := signature.GetMetadata()
+	engine.signaturesMutex.Lock()
+	defer engine.signaturesMutex.Unlock()
+	if engine.signatures[signature] != nil {
+		// signature already exists
+		return metadata.ID, nil
+	}
+	c := make(chan types.Event)
+	engine.signatures[signature] = c
+
+	// insert in engine.signaturesIndex map
+	for _, selectedEvent := range selectedEvents {
+		engine.signaturesIndex[selectedEvent] = append(engine.signaturesIndex[selectedEvent], signature)
+	}
+	go engine.signatureStartFunc(signature, c)
+	return metadata.ID, nil
+}
+
+//UnloadSignature will remove from Engine data structures the given signature and stop its handling goroutine
+func (engine *Engine) UnloadSignature(signature types.Signature) error {
+	selectedEvents, err := signature.GetSelectedEvents()
+	if err != nil {
+		return fmt.Errorf("failed to unload signature: %w", err)
+	}
+	metadata, err := signature.GetMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to unload signature: %w", err)
+	}
+	sigId := metadata.ID
+	engine.signaturesMutex.Lock()
+	defer engine.signaturesMutex.Unlock()
+
+	// remove from engine.signatures map
+	c := engine.signatures[signature]
+
+	// remove from engine.signaturesIndex map
+	for _, selectedEvent := range selectedEvents {
+		signatures := engine.signaturesIndex[selectedEvent]
+		for i, sig := range signatures {
+			metadata, _ := sig.GetMetadata()
+			if metadata.ID == sigId {
+				// signature found, remove it
+				signatures = append(signatures[:i], signatures[i+1:]...)
```

Comment:
This will panic if the signature is the last element in the slice (index i+1 is out of bounds).
As we don't care about the order of the signatures, you can do something like:
```
s[i] = s[len(s)-1]
s = s[:len(s)-1]
```

Reply:
I don't think it will panic. Note I am taking 'i+1'.
https://play.golang.org/p/5zyN-OUAe0K

Reply:
You are right.
Anyways, I think that the method I suggested will be more efficient, but not a must